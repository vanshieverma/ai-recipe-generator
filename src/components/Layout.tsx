import React from 'react';
import { useSession } from 'next-auth/react';
import { useRouter } from 'next/router';
import Header from './Header';
import Hero from '../pages/Hero';
import Loading from './Loading'
import ErrorPage from '../pages/auth/error';

/* Note all components will be wrapped in this component which in turn is rendered by _app.tsx */
const Layout = ({ children }: { children: React.ReactNode }) => {
  const { data: session, status } = useSession();
  const router = useRouter();
  const { error: signinError } = router.query;

  if (signinError) {
    return <ErrorPage />
  }

  if (router.pathname === '/_error') {
    return <ErrorPage message="Page not found"/>
  }

  if (status === 'loading') {
    return <Loading />;
  }

  if (!session) {
    return <Hero />
  }

  return (
    <div>
      <Header user={session.user} />
      <main className="min-h-screen bg-green-50">{children}</main>
    </div>
  );
};

export default Layout;
