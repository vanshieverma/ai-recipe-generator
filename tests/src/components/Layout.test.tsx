import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import { useSession } from "next-auth/react"
import { useRouter } from 'next/router';
import Layout from "../../../src/components/Layout";

jest.mock("next-auth/react",()=>({
    ...jest.requireActual("next-auth/react"),
    useSession: jest.fn(() => ({
        data: null,
        status: 'authenticated'
    }))
}))

jest.mock("next/router", () => ({
    useRouter: jest.fn(() => ({
        query: {},
        pathname:''
    })),
}))

jest.mock("../../../src/components/Header")

describe('The Layout', () => {
    it('renders the Error page for a sign in error', () => {
        (useRouter as jest.Mock).mockImplementationOnce(() => ({
            query: {
                error: 'a sign in error'
            }
        })).mockImplementationOnce(() => ({
            query: {
                error: 'a sign in error'
            }
        }))
        const { container } = render(
            <Layout><div>Child component</div>
            </Layout>)
        expect(screen.getByText('Sign In Error')).toBeInTheDocument();
        expect(screen.getByText('An unexpected error: "a sign in error" occurred. Please try again later.')).toBeInTheDocument();
    })

    it('renders the Error a page not found', () => {
        (useRouter as jest.Mock).mockImplementationOnce(() => ({
            query:{},
            pathname: '/_error'
        }))
        const { container } = render(
            <Layout><div>Child component</div>
            </Layout>)
        expect(screen.getByText('Page not found')).toBeInTheDocument();
    })

    it('renders the Loading wheel if session is loading', () => {
        (useSession as jest.Mock).mockImplementationOnce(() => ({
            data: null,
            status: "loading"
        }))
        const { container } = render(
            <Layout>
                <div>Child component</div>
            </Layout>)
        expect(container).toMatchSnapshot()
    })
    it('renders the hero if there is no session', () => {
        const { container } = render(
            <Layout><div>Child component</div>
            </Layout>)
        expect(container).toMatchSnapshot()
        expect(screen.getByText('Get started')).toBeInTheDocument();
    })
    it('renders the the children if session is valid', () => {
        (useSession as jest.Mock).mockImplementationOnce(() => ({
            data: {
                user: {
                    name: "mockuser",
                    image: "mockImage",
                    email: "mockEmail"
                },
            },
            status: "authenticated",
        }))
        const { container } = render(
            <Layout>
                <div>Child component</div>
            </Layout>)
        expect(screen.getByText('Child component')).toBeInTheDocument();
    })
})