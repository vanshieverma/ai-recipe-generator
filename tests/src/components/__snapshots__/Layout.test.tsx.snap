// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Layout renders the Loading wheel if session is loading 1`] = `
<div>
  <div
    class="flex items-center justify-center mt-5"
  >
    <div
      class="relative"
    >
      <div
        class="h-24 w-24 rounded-full border-t-8 border-b-8 border-gray-200"
      />
      <div
        class="absolute top-0 left-0 h-24 w-24 rounded-full border-t-8 border-b-8 border-blue-500 animate-spin"
      />
    </div>
  </div>
</div>
`;

exports[`The Layout renders the hero if there is no session 1`] = `
<div>
  <div
    class="bg-white"
  >
    <header
      class="absolute inset-x-0 top-0 z-50"
    >
      <nav
        aria-label="Global"
        class="flex items-center justify-between p-6 lg:px-8"
      >
        <div
          class="flex lg:flex-1"
        >
          <a
            class="-m-1.5 p-1.5"
            href="#"
          >
            <span
              class="sr-only"
            >
              Smart Recipe Generator
            </span>
            <img
              alt="Smart Recipe Generator Logo"
              data-nimg="1"
              decoding="async"
              height="75"
              loading="lazy"
              src="/logo.svg"
              style="color: transparent;"
              width="75"
            />
          </a>
        </div>
        <div
          class="flex lg:hidden"
        >
          <button
            class="-m-2.5 inline-flex items-center justify-center rounded-md p-2.5 text-gray-700"
            type="button"
          >
            <span
              class="sr-only"
            >
              Open main menu
            </span>
            <svg
              aria-hidden="true"
              class="h-6 w-6"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
        </div>
        <div
          class="hidden lg:flex lg:gap-x-12"
        >
          <button
            class="text-sm font-semibold leading-6 text-gray-900"
          >
            Product
          </button>
          <button
            class="text-sm font-semibold leading-6 text-gray-900"
          >
            Features
          </button>
          <button
            class="text-sm font-semibold leading-6 text-gray-900"
          >
            About
          </button>
        </div>
        <div
          class="hidden lg:flex lg:flex-1 lg:justify-end"
        >
          <button
            class="text-sm font-semibold leading-6 text-gray-900"
          >
            Log in With Google 
            <span
              aria-hidden="true"
            >
              →
            </span>
          </button>
        </div>
      </nav>
    </header>
    <div
      class="relative isolate px-6 pt-14 lg:px-8"
    >
      <div
        aria-hidden="true"
        class="absolute inset-x-0 -top-40 -z-10 transform-gpu overflow-hidden blur-3xl sm:-top-80"
      >
        <div
          class="relative left-[calc(50%-11rem)] aspect-[1155/678] w-[36.125rem] -translate-x-1/2 rotate-[30deg] bg-gradient-to-tr from-[#ff80b5] to-[#9089fc] opacity-30 sm:left-[calc(50%-30rem)] sm:w-[72.1875rem]"
          style="clip-path: polygon(74.1% 44.1%, 100% 61.6%, 97.5% 26.9%, 85.5% 0.1%, 80.7% 2%, 72.5% 32.5%, 60.2% 62.4%, 52.4% 68.1%, 47.5% 58.3%, 45.2% 34.5%, 27.5% 76.7%, 0.1% 64.9%, 17.9% 100%, 27.6% 76.8%, 76.1% 97.7%, 74.1% 44.1%);"
        />
      </div>
      <div
        class="mx-auto max-w-2xl py-32 sm:py-48 lg:py-56"
      >
        <div
          class="hidden sm:mb-8 sm:flex sm:justify-center"
        >
          <div
            class="relative rounded-full px-3 py-1 text-sm leading-6 text-gray-600 ring-1 ring-gray-900/10 hover:ring-gray-900/20"
          >
            Discover our new AI-powered recipe generator.
             
            <a
              class="font-semibold text-indigo-600"
              href="https://github.com/Dereje1/smart-recipe-generator"
            >
              <span
                aria-hidden="true"
                class="absolute inset-0"
              />
              Learn more 
              <span
                aria-hidden="true"
              >
                →
              </span>
            </a>
          </div>
        </div>
        <div
          class="text-center"
        >
          <h1
            class="text-4xl font-bold tracking-tight text-gray-900 sm:text-6xl"
          >
            Generate Delicious Recipes with Your Ingredients
          </h1>
          <p
            class="mt-6 text-lg leading-8 text-gray-600"
          >
            Simply input your available ingredients, select dietary preferences, and let our AI create unique and delicious recipes just for you.
          </p>
          <div
            class="mt-10 flex items-center justify-center gap-x-6"
          >
            <button
              class="rounded-md bg-indigo-600 px-3.5 py-2.5 text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
            >
              Get started
            </button>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="absolute inset-x-0 top-[calc(100%-13rem)] -z-10 transform-gpu overflow-hidden blur-3xl sm:top-[calc(100%-30rem)]"
      >
        <div
          class="relative left-[calc(50%+3rem)] aspect-[1155/678] w-[36.125rem] -translate-x-1/2 bg-gradient-to-tr from-[#ff80b5] to-[#9089fc] opacity-30 sm:left-[calc(50%+36rem)] sm:w-[72.1875rem]"
          style="clip-path: polygon(74.1% 44.1%, 100% 61.6%, 97.5% 26.9%, 85.5% 0.1%, 80.7% 2%, 72.5% 32.5%, 60.2% 62.4%, 52.4% 68.1%, 47.5% 58.3%, 45.2% 34.5%, 27.5% 76.7%, 0.1% 64.9%, 17.9% 100%, 27.6% 76.8%, 76.1% 97.7%, 74.1% 44.1%);"
        />
      </div>
    </div>
  </div>
</div>
`;
